src/async_wrap.cc

```diff
@@ -198,37 +200,110 @@ class PromiseWrap : public AsyncWrap {
     MakeWeak();
   }
 
+  PromiseWrap(Environment* env, Local<Object> object, double asyncId,
+    double triggerAsyncId)
+      : AsyncWrap(env, object, PROVIDER_PROMISE, asyncId, triggerAsyncId) {
+    MakeWeak();
+  }
+
   SET_NO_MEMORY_INFO()
   SET_MEMORY_INFO_NAME(PromiseWrap)
   SET_SELF_SIZE(PromiseWrap)
 
   static PromiseWrap* New(Environment* env,
                           Local<Promise> promise,
                           PromiseWrap* parent_wrap,
+                          bool isInit,
                           bool silent);
-  static void getIsChainedPromise(Local<String> property,
-                                  const PropertyCallbackInfo<Value>& info);
+  static void getAsyncId(Local<Name> property,
+                         const PropertyCallbackInfo<Value>& args);
+  static void getTriggerAsyncId(Local<Name> property,
+                                const PropertyCallbackInfo<Value>& args);
+
+  static void Initialize(Environment* env);
 };
 
 PromiseWrap* PromiseWrap::New(Environment* env,
                               Local<Promise> promise,
                               PromiseWrap* parent_wrap,
+                              bool isInit,
                               bool silent) {
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
   Local<Object> obj;
-  if (!env->promise_wrap_template()->NewInstance(env->context()).ToLocal(&obj))
+  if (!env->promise_wrap_template()->NewInstance(context).ToLocal(&obj))
     return nullptr;
+
   obj->SetInternalField(PromiseWrap::kIsChainedPromiseField,
-                        parent_wrap != nullptr ? v8::True(env->isolate())
-                                               : v8::False(env->isolate()));
+                        parent_wrap != nullptr ? v8::True(isolate)
+                                               : v8::False(isolate));
   CHECK_NULL(promise->GetAlignedPointerFromInternalField(0));
   promise->SetInternalField(0, obj);
+
+  // Avoid unless this is from an init event
```

Comment:
This comment is a bit misleading, as we avoid contents of this block when it's an `init` event. Does it make sense to change it to something like `Avoid in case of an init event`?

Reply:
Oh, yep. You're right.

Reply:
Fixed.